finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4DispatcherBase.scala

```diff
@@ -42,6 +50,33 @@ trait Netty4DispatcherBase[SendMsg <: Message, RecvMsg <: Message] {
     streams.size()
   }
 
+  @volatile private[this] var pingFramePromise: Promise[Unit] = null
+
+  protected[this] val OutstandingPingEx = Failure("Outstanding ping on HTTP/2 connection")
+  protected[this] val eventLoop = transport.context match {
+    case trans: HasExecutor => trans.executor
+    case _ => null
+  }
+
+  protected[this] def ping(): Future[Unit] = {
+    val done = new Promise[Unit]
+    eventLoop.execute(
+      new Runnable {
+        override def run(): Unit = {
+          if (pingFramePromise == null) {
+            writer.ping()
+            pingFramePromise = done
+          } else {
+            done.setException(OutstandingPingEx)
```

Comment:
possible race condition where `done` could be set to null before this is called

Reply:
Looking at this a little more, the `done` references the promise within the method and in this case, when we set the exception here, we are notifying the Failure detector that things went wrong.

Reply:
ah, you're right